site/app/entities/forum/Thread.php

```diff
@@ -72,9 +134,54 @@ class Thread {
     #[ORM\InverseJoinColumn(name: "category_id", referencedColumnName: "category_id")]
     protected Collection $categories;
 
+    /**
+     * @return Collection<Category>
+     */
+    public function getCategories(): Collection {
+        return $this->categories;
+    }
+
     /**
      * @var Collection<ThreadAccess>
      */
     #[ORM\OneToMany(mappedBy: "thread", targetEntity: ThreadAccess::class)]
     protected Collection $viewers;
+
+    /**
+     * @var Collection<StudentFavorite>
+     */
+    #[ORM\OneToMany(mappedBy: "thread", targetEntity: StudentFavorite::class)]
+    protected Collection $favorers;
+
+    public function isUnread(string $user_id): bool {
+        return !$this->getNewPosts($user_id)->isEmpty();
+    }
+
+    /**
+     * @return Collection<Post>
+     */
+    public function getNewPosts(string $user_id): Collection {
+        $last_view = $this->viewers->filter(function ($x) use ($user_id) {
+            return $user_id === $x->getUserId();
+        })->first();
+
+        if ($last_view === false) {
+            return $this->posts;
+        }
+        return $this->posts->filter(function ($x) use ($last_view) {
+            return $x->isUnread($last_view);
+        });
+    }
```

Comment:
Would it make sense to move these to a repository instead?  My understanding (which could very well be incorrect) is that methods which don't strictly relate to a given instance of an entity should reside in a repository.

By performing these filters in PHP, a lot of extra data is going to have to be queried.  If you moved these to a repository, the repository could add the right query filters such that only the necessary data is loaded.

Reply:
The repository can (and I configured it to) fetch the relevant entities when getting the thread, so all the data we need is in memory by the time any of these functions can get called. A lot of these methods are called in loops (ex. html classes depending on thread read/unread) and I think its simpler to fetch the necessary data with the thread than do a separate query.

In the case that calling these creates a problematic number of queries for some functionality down the line, the programmer can easily select and join the necessary data in the repository.